### PR TITLE
api: fix splice update operation

### LINKED ISSUE #348: bug: invalid delete and splice operations
https://www.tarantool.io/ru/doc/latest/dev_guide/internals/iproto/requests/#examples

`IPROTO_UPDATE` operation `#` (delete) could contains 2 fields, `:` (splice) - 5 fields. But we always use 3-fields format:

https://github.com/tarantool/go-tarantool/blob/c564e6d75d10b1bfef276ebacc2eb7cf4acfdb24/client_tools.go#L64-L69

### PR DESCRIPTION
**Pull request by DerekBum**
Splice update operation (`:`) accepts 5 and only 5 arguments. It was parsed and encoded incorrectly with only 3 argument (as every other update operations).
Also fixed the same operation for the crud.

`Op` struct made private.
Change all `Upsert` and `Update` requests to accept `*tarantool.Operations` as `ops` parameters instead of the `interface{}`.

I didn't forget about (remove if it is not applicable):

- [x] Tests (see [documentation](https://pkg.go.dev/testing) for a testing package)
- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)
- [x] Documentation (see [documentation](https://go.dev/blog/godoc) for documentation style guide)

Closes #348

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -80,6 +80,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
   `prefer_replica`, `balance`) setup for crud.GetRequest (#335)
 - Tests with crud 1.4.0 (#336)
 - Tests with case sensitive SQL (#341)
+- Splice update operation (#348). It now accepts 5 arguments as it should be.
```
Comment (oleg-jukovec) [inline CHANGELOG.md:83]:
```suggestion
- Splice update operation accepts 3 arguments instead of 5 (#348)
```

File: client_tools.go [hunk 2]
```diff
@@ -140,21 +171,3 @@ func (ops *Operations) Delete(field int, arg interface{}) *Operations {
 func (ops *Operations) Assign(field int, arg interface{}) *Operations {
 	return ops.append(assignOperator, field, arg)
 }
-
-type OpSplice struct {
```
Comment (oleg-jukovec) [inline client_tools.go:144]: Please add a note to README.md into the migration article about the change in the public API.

**Review by oleg-jukovec — COMMENTED** [review]
Let's add examples of using the type Operations and []Op{} as update arguments.

(I see ExampleUpdateRequest, but that's not quite it).

File: crud/operations.go [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package crud
 
+import "github.com/vmihailenco/msgpack/v5"
```
Comment (oleg-jukovec) [inline crud/operations.go:3]: I prefer that style in the repository (if this is not done by the autoformatter).
```suggestion
import (
    "github.com/vmihailenco/msgpack/v5"
)
```

File: CHANGELOG.md (hunk 1 above)
Comment (DerekBum) [reply CHANGELOG.md]: Updated.

File: crud/operations.go (hunk 3 above)
Comment (DerekBum) [reply crud/operations.go]: Updated.

File: README.md [hunk 4]
```diff
@@ -253,6 +253,13 @@ interface to get information if the usage of space and index names in requests
 is supported.
 * `Schema` structure no longer implements `SchemaResolver` interface.
 
+#### Client tools changes
+
+* `Op` struct made private.
+* Remove `OpSplice` struct.
+* `Operations` struct made an instance of `msgpack.CustomEncoder` interface.
```
Comment (oleg-jukovec) [inline README.md:260]: You don't need to notion this, it doesn't break anything.

File: client_tools.go [hunk 5]
```diff
@@ -82,17 +108,26 @@ const (
 
 // Operations is a collection of update operations.
 type Operations struct {
-	ops []Op
+	ops []operation
+}
+
```
Comment (oleg-jukovec) [inline client_tools.go:113]: A comment is missed.

File: request.go [hunk 6]
```diff
@@ -1268,7 +1267,7 @@ func NewUpsertRequest(space interface{}) *UpsertRequest {
 	req.rtype = iproto.IPROTO_UPSERT
 	req.setSpace(space)
 	req.tuple = []interface{}{}
-	req.ops = []interface{}{}
+	req.ops = NewOperations()
```
Comment (oleg-jukovec) [inline request.go:1270]: Let's not create an empty object here, this leads to unnecessary allocation to the request.

It's better to check `req.ops == nil` during encoding and use []interface{}{} instead if so.

File: request.go [hunk 7]
```diff
@@ -1202,7 +1203,7 @@ func NewUpdateRequest(space interface{}) *UpdateRequest {
 	req.rtype = iproto.IPROTO_UPDATE
 	req.setSpace(space)
 	req.key = []interface{}{}
-	req.ops = []interface{}{}
+	req.ops = NewOperations()
```
Comment (oleg-jukovec) [inline request.go:1206]: Let's not create an empty object here, this leads to unnecessary allocation to the request.

It's better to check `req.ops == nil` during encoding and use []interface{}{} instead if so.

File: crud/operations.go [hunk 8]
```diff
@@ -23,11 +27,43 @@ const (
...
 	Operator Operator
 	Field    interface{} // Number or string.
 	Value    interface{}
+	// Pos, Len, Replace fields used in the Splice operation.
+	Pos     int
+	Len     int
+	Replace string
+}
+
+func (o Operation) EncodeMsgpack(enc *msgpack.Encoder) error {
```
Comment (oleg-jukovec) [inline crud/operations.go:42]: Missed comment.

File: client_tools.go [hunk 9]
```diff
@@ -82,17 +108,26 @@ const (
...
+func (ops *Operations) EncodeMsgpack(enc *msgpack.Encoder) error {
+	return enc.Encode(ops.ops)
 }
```
Comment (oleg-jukovec) [inline client_tools.go:116]: Let's add a unit-test for the new public method.

File: crud/tarantool_test.go [hunk 10]
```diff
@@ -40,10 +40,62 @@ var startOpts test_helpers.StartOpts = test_helpers.StartOpts{
 var timeout = float64(1.1)
 
 var operations = []crud.Operation{
+	// Insert new fields,
+	// because double update of the same field results in an error.
 	{
-		Operator: crud.Assign,
```
Comment (DifferentialOrange) [inline crud/tarantool_test.go:44]: We don't test assign now? It's the most useful update operation.

File: test_helpers/utils.go [hunk 11]
```diff
@@ -200,6 +200,14 @@ func SkipIfWatchOnceUnsupported(t *testing.T) {
 	SkipIfFeatureUnsupported(t, "watch once", 3, 0, 0)
 }
 
+// SkipIfCrudSpliceUnsupported skips test run if Tarantool without support of
+// crud splice update operation is used.
```
Comment (DifferentialOrange) [inline test_helpers/utils.go:204]: crud splice is core Tarantool splice

Comment (DerekBum) [reply test_helpers/utils.go:204]: But for some reason, using splice in crud is not supported for Tarantool < 2.0. While we can use splice from the core.

File: README.md (hunk 4 above)
Comment (DerekBum) [reply README.md]: Updated.

File: client_tools.go (hunk 5 above)
Comment (DerekBum) [reply client_tools.go:113]: Updated.

File: request.go (hunk 6 above)
Comment (DerekBum) [reply request.go]: Updated.

File: request.go (hunk 7 above)
Comment (DerekBum) [reply request.go]: Updated.

File: crud/operations.go (hunk 8 above)
Comment (DerekBum) [reply crud/operations.go]: Updated.

File: client_tools.go (hunk 9 above)
Comment (DerekBum) [reply client_tools.go:117]: New unit-test added to the `client_tools_test.go`.

File: crud/tarantool_test.go (hunk 10 above)
Comment (DerekBum) [reply crud/tarantool_test.go:44]: Oops, restored the assign test.

File: client_tools.go [hunk 12]
```diff
@@ -82,17 +108,27 @@ const (
...
+
+// EncodeMsgpack encodes Operations as an array of operations.
+func (ops *Operations) EncodeMsgpack(enc *msgpack.Encoder) error {
+	return enc.Encode(ops.ops)
 }
 
 // NewOperations returns a new empty collection of update operations.
 func NewOperations() *Operations {
-	ops := new(Operations)
+	ops := &Operations{[]operation{}}
```
Comment (oleg-jukovec) [inline client_tools.go:121]:
```suggestion
	return &Operations{[]operation{}}
```

File: client_tools_test.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+		BitwiseOr(1, 2).
+		BitwiseXor(1, 2).
+		Splice(1, 2, 3, "a").
+		Insert(1, 2).
+		Delete(1, 2).
+		Assign(1, 2)
+
+	refBuf := []byte{153, 147, 161, 43, 1, 2, 147, 161, 45, 1, 2, 147, 161, 38,
+		1, 2, 147, 161, 124, 1, 2, 147, 161, 94, 1, 2, 149, 161, 58, 1, 2, 3,
+		161, 97, 147, 161, 33, 1, 2, 147, 161, 35, 1, 2, 147, 161, 61, 1, 2}
```
Comment (oleg-jukovec) [inline client_tools_test.go:26]: Let's comment here at least on the beginning of each operation.

File: crud/operations.go [hunk 14]
```diff
@@ -23,11 +27,44 @@ const (
...
 	Field    interface{} // Number or string.
 	Value    interface{}
+	// Pos, Len, Replace fields used in the Splice operation.
+	Pos     int
+	Len     int
+	Replace string
+}
+
+// EncodeMsgpack encodes Operation.
+func (o Operation) EncodeMsgpack(enc *msgpack.Encoder) error {
```
Comment (oleg-jukovec) [inline crud/operations.go:43]: Let's test it with a unit-test too.

**Review by oleg-jukovec — APPROVED** [review]
Thank you for the patch! Just few notes.

File: client_tools.go (hunk 12 above)
Comment (DerekBum) [reply client_tools.go]: Updated (left `return new(Operations)`).

File: client_tools_test.go (hunk 13 above)
Comment (DerekBum) [reply client_tools_test.go]: Updated.

File: crud/operations.go (hunk 14 above)
Comment (DerekBum) [reply crud/operations.go:43]: Added a new test to the `crud/operations_test.go`.

File: README.md [hunk 15]
```diff
@@ -253,6 +253,12 @@ interface to get information if the usage of space and index names in requests
 is supported.
 * `Schema` structure no longer implements `SchemaResolver` interface.
 
+#### Client tools changes
+
+* `Op` struct made private.
+* Remove `OpSplice` struct.
+* `Operations.Splice` method now accepts 5 arguments instead of 3.
```
Comment (oleg-jukovec) [inline README.md:260]: This changes more about requests.
Let's also hightlight that Update/Upsert/UpdateAsync... API changed.

**Review by oleg-jukovec — APPROVED** [review]

File: client_tools.go (hunk 12 above)
Comment (oleg-jukovec) [reply client_tools.go]: Does encoder.Encode(NewOperations()) work fine? (encodes an empty msgpack array).

Comment (DerekBum) [reply client_tools.go]: If we want the result of `Encode(NewOperation())` be equal to `Encode([]interface{}{})`, then we need to use `&Operations{[]operation{}}`. Updated.

File: README.md (hunk 15 above)
Comment (DerekBum) [reply README.md]: `README` updated.

File: README.md [hunk 16]
```diff
@@ -229,9 +229,17 @@ of the requests is an array instead of array of arrays.
...
 IPROTO constants have been moved to a separate package [go-iproto](https://github.com/tarantool/go-iproto).
 
-#### Request interface
+#### Request changes
 
 * The method `Code() uint32` replaced by the `Type() iproto.Type`.
+* `Op` struct for update operations made private.
+* Removed `OpSplice` struct.
+* `Operations.Splice` method now accepts 5 arguments instead of 3.
+* Requests `Update`, `UpdateAsync`, `UpdateTyped`, `Upsert`, `UpsertAsync` no
```
Comment (oleg-jukovec) [inline README.md:238]: Update/UpsertTyped?

File: test_helpers/utils.go (hunk 11 above)
Comment (DifferentialOrange) [reply test_helpers/utils.go:204]: Can you show the exact error?

Comment (DerekBum) [reply test_helpers/utils.go:204]: Sure. For example:
https://github.com/tarantool/go-tarantool/actions/runs/6875782438/job/18700075402

```
    --- FAIL: TestCrudProcessData/Update (0.00s)
        tarantool_test.go:583: Failed to perform CRUD request on CRUD side: map[interface {}]interface {}{"class_name":"UpdateError", "err":"Failed to update: Unknown UPDATE operation", "file":"...ool/crud/testdata/.rocks/share/tarantool/crud/update.lua", "line":0xb0, "str":"UpdateError: Failed to update: Unknown UPDATE operation"}
```

But for the core:
```
PASS
ok  	github.com/tarantool/go-tarantool/v2	4.867s
```

Comment (DifferentialOrange) [reply test_helpers/utils.go:204]: It seems that this one is more "broken" than "unsupported": https://github.com/tarantool/crud/issues/397

Thank for the report. I think it's worth to add a link to the issue to skip helper.

**Review by DifferentialOrange — APPROVED** [review]
Thank you for the fix! No objections from my side after resolving the last discussion.

File: test_helpers/utils.go (hunk 11 above)
Comment (DerekBum) [reply test_helpers/utils.go]: Updated function name and added a link to an issue.

**Pull request merged**